### PR TITLE
Implement `aten::_softmax` | feat(torchlib)

### PR DESCRIPTION
**Pull request by justinchuby**
Move `aten_special_softmax` -> `aten_softmax`, support alias and when dtype input is None. Implement `_softmax` using `softmax`.

An alternative would be to create a decomp in pytorch from _softmax to softmax. But since pytorch optimization tends to create decomp to private ops (and not the other way around) in different optimizations steps as @BowenBao pointed out, we still need to support _softmax here.

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -67,6 +67,29 @@ def aten__local_scalar_dense_int(self: IntType) -> INT64:
...
+    # trace_only because we need to cast conditionally based on half_to_float
+    if half_to_float:
+        self = op.Cast(self, to=FLOAT.dtype)
+
+    return aten_softmax_no_dtype(self, dim)
+
+
+@torch_op("aten::_softmax", trace_only=True)
+def aten__softmax(
+    self: TFloatHighPrecision, dim: int, half_to_float: bool
```
Comment (fatcat-z) [inline onnxscript/function_libs/torch_lib/ops/core.py:83]: What does these 2 _softmax() functions has different dtypes for self?

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py [hunk 2]
```diff
@@ -1197,7 +1197,7 @@ def _where_input_wrangler(
     TorchLibOpInfo("sinh", core_ops.aten_sinh),
     TorchLibOpInfo(
         "softmax",
-        special_ops.aten_special_softmax,
+        core_ops.aten_softmax,
```
Comment (fatcat-z) [inline onnxscript/tests/function_libs/torch_lib/ops_test_data.py:1200]: Should we add aten__softmax into test as well?

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/core.py:83]: This is because we want to cast float16 to float32 when half_to_float is true. Since we have no idea what the input dtype will be within the function body, we rely on the dispatcher to pick the function that already limits the input dtype it can accept to know we are dealing with float16 types.

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_data.py:1200]: We should

Comment (fatcat-z) [reply onnxscript/tests/function_libs/torch_lib/ops_test_data.py:1200]: > We should

Will it be done in this PR or next one in future?

Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_data.py:1200]: I will update the PR. Sorry for the confusion.

**Review by fatcat-z — APPROVED** [review]
